sigstore/sign.py

```diff
@@ -229,6 +230,45 @@ def sign_dsse(
 
         return self._finalize_sign(cert, content, proposed_entry)
 
+    def sign_dsse_envelope(
+            self,
+            envelope: dsse.Envelope,
+    ) -> Bundle:
+        """
+        Signs the provided envelope's payload, and returns a
+        `Bundle containing the signed envelope and the verification
+        material.
+
+        Args:
+            envelope (dsse.Envelope): The envelope to be signed.
+
+        Returns:
+            Bundle: The bundle containing the signed DSSE envelope.
+        """
+        cert = self._signing_cert()
+
+        b64_cert = base64.b64encode(
+            cert.public_bytes(encoding=serialization.Encoding.PEM)
+        )
+        envelope = dsse._sign_envelope(
+            self._private_key,
+            envelope,
+        )
+
+        proposed_entry = rekor_types.Dsse(
+            spec=rekor_types.dsse.DsseSchema(
+                # NOTE: mypy can't see that this kwarg is correct due to two interacting
+                # behaviors/bugs (one pydantic, one datamodel-codegen):
+                # See: <https://github.com/pydantic/pydantic/discussions/7418#discussioncomment-9024927>
+                # See: <https://github.com/koxudaxi/datamodel-code-generator/issues/1903>
+                proposed_content=rekor_types.dsse.ProposedContent(  # type: ignore[call-arg]
+                    envelope=envelope.to_json(),
+                    verifiers=[b64_cert.decode()],
+                ),
+            ),
+        )
+        return self._finalize_sign(cert, envelope, proposed_entry)
```

Comment:
Making sure I'm not missing anything: does this actually work yet? My understanding was that Rekor doesn't yet support arbitrary DSSE envelopes, since it expects a valid in-toto statement within the envelope.

CC @haydentherapper

Reply:
That is correct, https://github.com/sigstore/rekor/blob/main/pkg/types/dsse/v0.0.1/entry.go#L112-L146, also pointed out in https://github.com/sigstore/sigstore-python/issues/982#issuecomment-2108913284.

The reason is so that Rekor can extract index keys from the statement. We can add other payload types.

Reply:
Thanks @haydentherapper! I was feeling some deja vu at this 😅 

Given the above, I'm not sure we can merge this/it makes sense to merge until Rekor supports other payload types (since users who attempt to use it will hit a hard error and think that sigstore-python is broken).

(Another option here would be to use a DSSE envelope with an arbitrary payload, but to emit a `hashedrekord` entry instead. I have no strong opinion on whether this is a good idea or not, but it'll likely cause a lot of incompatibilities with the other clients in the ecosystem.)

Reply:
Oh ok, I missed that. Sorry!
We could log a warning when an arbitrary DSSE envelope is used and skip the Rekor  post in finalize sign.
Would that make sense or would it mess up the verification itself?

Reply:
> Would that make sense or would it mess up the verification itself?

That would mess up the verification itself, unfortunately -- that would result in a bundle with no transparency log entries, which would both be invalid and also incompatible with other clients.

Reply:
Even if we made index keys optional for unparseable DSSEs, this would be a breaking change for log monitors that expect an intoto payload in order to extract subject values.

Reply:
How could we proceed with this then?

* Would it be possible to use the SHA256 of the payload as index key for arbitrary payloads?
* Should we define our DSSE envelope payload type before and then add it to rekor?  <- this wouldn't be a solution for arbitrary payloads but helps the model signing effort.

Reply:
@haydentherapper I meant to bring your focus back to this by re-triggering the review ;)